src/test.rs

```diff
@@ -394,6 +395,70 @@ fn layer2() {
 }
 
 #[test]
+fn layer2_timeouts() {
+    use std::time::Duration;
+    use std::io::ErrorKind;
+    use datalink;
+    use datalink::Channel::Ethernet;
+    use packet::ethernet::{EtherTypes, MutableEthernetPacket};
+
+    const ETHERNET_HEADER_LEN: usize = 14;
+
+    let interface = get_test_interface();
+
+    let mut packet = [0u8; ETHERNET_HEADER_LEN + IPV4_HEADER_LEN + UDP_HEADER_LEN + TEST_DATA_LEN];
+
+    {
+        let mut ethernet_header = MutableEthernetPacket::new(&mut packet[..]).unwrap();
+        ethernet_header.set_source(interface.mac_address());
+        ethernet_header.set_destination(interface.mac_address());
+        ethernet_header.set_ethertype(EtherTypes::Ipv4);
+    }
+
+    build_udp4_packet(&mut packet[..],
+                      ETHERNET_HEADER_LEN as usize,
+                      "l2tt",
+                      Some(&interface));
+
+    let (tx, rx) = channel();
+
+    let cfg = datalink::Config {
+        write_buffer_size: 4096,
+        read_buffer_size: 4096,
+        channel_type: datalink::ChannelType::Layer2,
+        bpf_fd_attempts: 100,
+        read_timeout: Some(Duration::from_millis(300)),
+        write_timeout: Some(Duration::from_millis(100))
+    };
+    let dlc = datalink::channel(&interface, &cfg);
+    let (_, mut dlrx) = match dlc {
+        Ok(Ethernet(tx, rx)) => (tx, rx),
+        Ok(_) => panic!("layer2_timeouts: unexpected L2 packet type"),
+        Err(e) => panic!("layer2_timeouts: unable to create channel: {}", e),
+    };
+
+    let res = thread::spawn(move || {
+        tx.send(()).unwrap();
+        let mut iter = dlrx.iter();
+        loop {
+            match iter.next() {
+                Ok(_) => {
+                    panic!("layer2_timeouts: should have exceeded timeout");
+                },
+                Err(e) => {
+                    assert!(e.kind() == ErrorKind::WouldBlock);
+                }
+            }
+        }
+    });
+    rx.recv().unwrap();
+    // Wait a while
+    thread::sleep(Duration::from_millis(1000));
+    // don't even send anything
```

Comment:
If Travis uses a recent-enough (works for me on Ubuntu 14.04) Linux, we could have Travis create an `iproute2` dummy interface for testing.